### PR TITLE
Credential template builder package

### PR DESCRIPTION
**Pull request by azdagron**
This package consolidates all of the CA and SVID template building logic into one spot where the Credential Composers can be applied.

The package does perform some input validation of parameters but does unfortunately does not perform validation against alterations returned by credential composers. This is because it is not possible to know how the extra extensions will impact the certificates until after they have passed through CreateCertificate, and we don't want to sign a credential after each CredentialComposer invocation.

As such, the validation will be left to the server CA layer after signing has taken place.

### CODE REVIEW

File: pkg/server/credtemplate/builder.go [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+			"iat": jwt.NewNumericDate(now),
+		},
+	}
+	if b.config.JWTIssuer != "" {
+		attributes.Claims["iss"] = b.config.JWTIssuer
+	}
+
+	for _, cc := range b.config.CredentialComposers {
+		var err error
+		attributes, err = cc.ComposeWorkloadJWTSVID(ctx, params.SPIFFEID, attributes)
```
Comment (amartinezfayo) [inline pkg/server/credtemplate/builder.go:336]: Should we detect and log a warning if a CredentialComposer plugin overwrites an attributes that was set by other CredentialComposer?

File: pkg/server/credtemplate/builder.go [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+
+	return tmpl, nil
+}
+
+func (b *Builder) buildX509SVIDTemplate(spiffeID spiffeid.ID, publicKey crypto.PublicKey, parentChain []*x509.Certificate, subject pkix.Name, ttl time.Duration) (*x509.Certificate, error) {
+	if len(parentChain) == 0 {
+		return nil, errors.New("parent chain required to build X509-SVID template")
+	}
+	if spiffeID.IsZero() {
+		return nil, fmt.Errorf("invalid X509-SVID ID: cannot be empty")
```
Comment (amartinezfayo) [inline pkg/server/credtemplate/builder.go:364]:
```suggestion
		return nil, errors.New("invalid X509-SVID ID: cannot be empty")
```

File: pkg/server/credtemplate/builder_test.go [hunk 3]
```diff
@@ -0,0 +1,1188 @@
...
+
+func TestNewBuilderSetsDefaults(t *testing.T) {
+	builder, err := credtemplate.NewBuilder(credtemplate.Config{
+		TrustDomain: td,
+	})
+	require.NoError(t, err)
+
+	config := builder.Config()
+
+	// Assert that the Clock and NewSerialNumber and then set them to nil
```
Comment (amartinezfayo) [inline pkg/server/credtemplate/builder_test.go:67]:
```suggestion
	// Assert that the Clock and NewSerialNumber are not nil and then set them to nil
```

File: pkg/server/credtemplate/builder_test.go [hunk 4]
```diff
@@ -0,0 +1,1188 @@
...
+		JWTSVIDTTL:      3 * time.Minute,
+		JWTIssuer:       "ISSUER",
+		AgentSVIDTTL:    4 * time.Minute,
+	}
+	builder, err := credtemplate.NewBuilder(configIn)
+	require.NoError(t, err)
+
+	configOut := builder.Config()
+
+	// Assert that the Clock and NewSerialNumber and then set them to nil
```
Comment (amartinezfayo) [inline pkg/server/credtemplate/builder_test.go:104]:
```suggestion
	// Assert that the Clock and NewSerialNumber are not nil and then set them to nil
```

**Review by amartinezfayo — COMMENTED** [review]
I've done a first pass. My main question is around the handling of multiple composers overwriting the same attribute.

File: pkg/server/credtemplate/builder.go (hunk 1 above)
Comment (azdagron) [reply pkg/server/credtemplate/builder.go:336]: I don't know. I could imagine situations where that might be useful, but they are contrived. Honestly, I don't know how people are going to use these plugins in the wild.

Comment (amartinezfayo) [reply pkg/server/credtemplate/builder.go:336]: I also agree that's hard to tell until we gather information about how this is used.

**Review by amartinezfayo — APPROVED** [review]
LGTM!

File: pkg/server/credtemplate/builder_test.go [hunk 5]
```diff
@@ -0,0 +1,1188 @@
...
+				config.CredentialComposers = []credentialcomposer.CredentialComposer{fakeCC{id: 1}}
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.Subject.CommonName = "OVERRIDE-1"
+				expected.PolicyIdentifiers = []asn1.ObjectIdentifier{{1}}
+				expected.ExtraExtensions = []pkix.Extension{{Id: makeOID(1), Value: []byte{1}}}
+			},
+		},
+		{
+			desc: "two composers",
```
Comment (MarcosDY) [inline pkg/server/credtemplate/builder_test.go:171]: may we set something on the first composer, to be sure it survives from the second one? actual test case is basically overriding all, so we are not sure if the first one ran. (the same comment is for all another tests that do something similar)

File: pkg/server/credtemplate/builder_test.go [hunk 6]
```diff
@@ -0,0 +1,1188 @@
...
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.Subject.CommonName = "OVERRIDE-1"
+				expected.DNSNames = []string{"OVERRIDE-1"}
+				expected.ExtraExtensions = []pkix.Extension{{Id: makeOID(1), Value: []byte{1}}}
+			},
+		},
+		{
+			desc: "two composers",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.CredentialComposers = []credentialcomposer.CredentialComposer{fakeCC{id: 2}, fakeCC{id: 2}}
```
Comment (MarcosDY) [inline pkg/server/credtemplate/builder_test.go:529]: IDs are the same for both composers (same in all `two composers` test cases

File: pkg/server/credtemplate/builder_test.go [hunk 7]
```diff
@@ -0,0 +1,1188 @@
...
+				params.DNSNames = []string{"DNSNAME1", "DNSNAME2"}
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.DNSNames = []string{"DNSNAME1", "DNSNAME2"}
+				// CommonName is set to first DNS name by default
+				expected.Subject.CommonName = "DNSNAME1"
+			},
+		},
+		{
+			desc: "with DNS names and subject",
```
Comment (MarcosDY) [inline pkg/server/credtemplate/builder_test.go:843]: maybe an overkill... 
but how about a test case with DNS and CommonName override? to be sure plugin wins

File: pkg/server/credtemplate/builder_test.go [hunk 8]
```diff
@@ -0,0 +1,1188 @@
...
+			expectErr: "invalid JWT-SVID audience: cannot be empty",
+		},
+		{
+			desc: "empty audience value otherwise ignored",
+			overrideParams: func(params *credtemplate.WorkloadJWTSVIDParams) {
+				params.Audience = []string{"", "AUDIENCE"}
+			},
+		},
+		{
+			desc: "multipler audience value",
```
Comment (MarcosDY) [inline pkg/server/credtemplate/builder_test.go:996]: s/multipler/multipler/g

**Review by MarcosDY — COMMENTED** [review]
Looks great!!

File: pkg/server/credtemplate/builder.go [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+	DefaultJWTSVIDTTL = time.Minute * 5
+
+	// NotBeforeCushion is how much of a cushion to subtract from the current
+	// time when determining the notBefore field of certificates to account
+	// for clock skew.
+	NotBeforeCushion = 10 * time.Second
+)
+
+// DefaultX509CASubject is the default subject set on workload X509SVIDs
+// TODO: this is a hostoric, but poor default. we should revisit.
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/server/credtemplate/builder.go]: added an issue and updated comment.

File: pkg/server/credtemplate/builder.go [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+
+	tmpl, err := b.buildX509SVIDTemplate(params.SPIFFEID, params.PublicKey, params.ParentChain, subject, params.TTL)
+	if err != nil {
+		return nil, err
+	}
+
+	// The first DNS name is also added as the CN by default. This happens
+	// even if the subject is provided explicitly in the params for backwards
+	// compatibility. Ideally we wouldn't do override the subject in this
+	// case. It is still overridable via the credential composers however.
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/server/credtemplate/builder.go:284]: I agree this needs to be clearly documented... do you have an alternate suggestion on how this is implemented that retains backwards compatibility?

In my mind, the CredentialComposer should be allowed to do anything, as long as it produces a SPIFFE spec conforming certificate.

File: pkg/server/credtemplate/builder_test.go (hunk 5 above)
Comment (azdagron) [reply pkg/server/credtemplate/builder_test.go:171]: Let me see what I can do...

File: pkg/server/credtemplate/builder_test.go (hunk 6 above)
Comment (azdagron) [reply pkg/server/credtemplate/builder_test.go:529]: Yikes. Thanks.

File: pkg/server/credtemplate/builder.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (MarcosDY) [reply pkg/server/credtemplate/builder.go:284]: agree, I think we must allow them to update it, so I dont see this as an issue but a feature,
but what worry me is than someone may be confused about possible results

**Review by MarcosDY — APPROVED** [review]
LGTM!!!

**Pull request merged**